Automate comparison of snippet output

anthonygalea opened the pull request:
> This pull request is not intended to be merged in its current state.

* **Goal**: Address current pain point of having to manually compare snippet output when testing after having upgraded processing/p5.
* **Ideally**: We would have a solution that could cover both clj and cljs.
* **Approach so far**: First looked into [kamera](https://github.com/oliyh/kamera) which is using [imagemagick](https://imagemagick.org) but opted instead to start simply using quil's own `save` function and then compare expected with actual output.
* **Snippets skipped**: Some snippets cannot be tested using this approach because they generate different output on each run. I'm just skipping the assertion in such cases. Any other ideas?
* **Related**: I don't see any CI server reference on the readme. Is it just the reference that's missing? If not does it make sense to invest some time adding some CI? I'm thinking something like travis or circleci.

nbeloglazov commented:
Wow, that's great start! Couple of questions/thoughts:

1. Running this test on different machines/os will likely result in subtly different images. So we'll need to have a way to set threshold image difference upon which test fails. I suspect kamera has some utilities for calculating such diffs between images (or maybe there are standalone clojure libraries for doing that).

2. We might end up adding prefixes to images depending on environment they run. For example add OS name if rendering differs significantly on different OSes. That's how we do image based browser tests at work: each image contains browser + OS in its name so image on chrome differs from image on firefox. But I think we could start without suffixes and see how it goes. 

3. Skipped snippets to due randomnes. I can't think of a good solution. We might just keep them skipped. 

4. CI for Quil. There is no CI setup. Because current test (run all snippets) doesn't really check much beside that it doesn't throw error I didn't bother setting up CI. But yes, with image tests we should have one. One thing is Quil tests require some graphics things and it might require more setup on CI part? I assume CI usually assumes there is no UI and runs them on display-less servers. But then I'm sure CI support browser tests with screenshots so there should an option to have CI with display. I'm ok with either Travis or CircleCI as long as it has display thing for Quil.

anthonygalea commented:
Thanks for your feedback. Some minor things which came up along the way:

* The 1 second pause in the snippet tests causes a problem with snippets `redraw` and `blend`. They both have `no-loop` in `setup` and when they render fast enough I think `q/exit` is never called. Does it make sense to say that if the developer wants to look at the output he should use the `manual` option? This would resolve the issue and also make the build faster.

* Also ran into an issue with `ortho-perspective` snippet locally. It seems to be related to picking a frame-rate of 1 because when I remove this I don't see the issue anymore. This is an excerpt from what I see in the console: `
finishLifecycleAction(com.jogamp.opengl.util.FPSAnimator$3): ++++++ timeout reached ++++++ main-FPSAWTAnimator#15-Timer30`. While I like the comparison on each frame between ortho and perspective I chose to adapt the snippet to render both side by side statically as it makes more sense anyway for automated testing. What do you think?

* Regarding CI, after experimenting locally settled on `image: circleci/clojure:lein-2.8.3-browsers`. If I understood correctly this image uses [xvfb](https://github.com/CircleCI-Public/circleci-dockerfiles/blob/master/clojure/images/lein-2.8.1/browsers/Dockerfile#L63) which allows us to follow [guidelines from processing for running without a display](https://github.com/processing/processing/wiki/Running-without-a-Display). You can see the output [here](https://circleci.com/gh/anthonygalea/quil/2) as well as run build locally if you want:
```bash
brew install circleci #or equivalent if on different os
circleci local execute --job build
```

* The only remaining issue with regards to CI is with opengl. Still need to look into this:
```
 actual: com.jogamp.opengl.GLException: Profiles [GL4bc, GL4, GL3bc, GL3, GLES3, GL2, GLES2] not available on device X11GraphicsDevice[type .x11, connection :99, unitID 0, handle 0x0, owner false, ResourceToolkitLock[obj 0x590cdf9, isOwner false, <6527a4eb, 5b22112a>[count 0, qsz 0, owner <NULL>]]]
 at com.jogamp.opengl.GLProfile.get (GLProfile.java:1039)
    com.jogamp.opengl.GLProfile.get (GLProfile.java:1050)
```

* Regarding threshold for comparison I will see what is possible with kamera.

anthonygalea commented:
The comparison with imagemagick seems to work locally. Tested against an upgrade to use processing 3.5.3.

The dockerfile for the container it runs on is:
```dockerfile
FROM circleci/clojure:lein-2.8.3-browsers

RUN sudo apt-get update && sudo apt-get install -y imagemagick
```
What I still haven't figured out is what to do about the opengl exceptions when running in the container. Any ideas?

nbeloglazov commented:
> Does it make sense to say that if the developer wants to look at the output he should use the manual option? 
You mean run manually on certain snippets like `no-loop` and `blend`? Yes it would be ok. 

> While I like the comparison on each frame between ortho and perspective I chose to adapt the snippet to render both side by side statically as it makes more sense anyway for automated testing. What do you think?
Makes sense. I'd probably even suggest to split them into two different snippets for testing. Actually I think we might end up having some snippets to be marked only for testing and some only for showing on quil.info. For example for ortho/perspectiive as example it's nice to have flip but for testing it doesn't make sense as you found out. I suspect we are going to have more such cases. 

> What I still haven't figured out is what to do about the opengl exceptions when running in the container. Any ideas?
No :( I'm ok with partial support for clj. And then having better support for cljs (web). Hopefully when running in browser webgl works better on CI compared to opengl.

anthonygalea commented:
Since this is useful as it is for local development I dropped CI for now and cleaned things up a bit so it can be merged as it is. Also added automated cljs test using [etaoin](https://github.com/igrishaev/etaoin) with firefox:
![eRIBEHhE6l](https://user-images.githubusercontent.com/1301852/57788779-f93de300-7737-11e9-93a4-1de8cf348134.gif)
You can run it using:
```
lein with-profile cljs-testing do cljsbuild once tests, ring server
```
and then in separate shell:
```
lein test quil.snippet :manual
```
Unfortunately the ability to [take screenshots of specific elements](https://github.com/igrishaev/etaoin#screening-elements) is currently limited to firefox but eventually this test could be expanded to include other browsers.

Also it requires geckodriver installed locally:
```
brew install geckodriver
```

nbeloglazov commented on src/cljc/quil/snippets/lights_camera/camera.cljc:
```diff
@@ -75,23 +75,27 @@
-  (q/camera 300 0 300 100 0 0 0 0 -1)
-  (q/fill 0 127 127)
-  (comment "draw 3 boxes with x coordinates 0, 100 and 200")
+  (q/perspective)
```
I'm confused a little bit. How can you have both ortho and perspective cameras in the same image?

anthonygalea replied:
My guess is also that it is not intended to be used that way. In p5 docs both calls to [perspective](https://p5js.org/reference/#/p5/perspective) and [ortho](https://p5js.org/reference/#/p5/ortho) are made in `setup`. But since it's not a must we can display both side by side. I was hoping fill is enough to distinguish but if you think this is anyway confusing perhaps it makes sense to have 3 snippets as you suggested earlier?
* 1 with static image showing ortho (for automated tests)
* 1 with static image showing perspective (for automated tests)
* 1 with dynamic image showing switch between perspective and ortho (for use on quil.info)

nbeloglazov replied:
But fill actually removes all  (q/ortho) boxes that were drawn before that, right? Yes, I think having two separate snippets will help to avoid the issue entirely.

anthonygalea replied:
No, just checked to be sure, both clj and cljs look like this:
<img width="568" alt="Screenshot 2019-05-18 at 10 33 33" src="https://user-images.githubusercontent.com/1301852/57966930-c888c400-7958-11e9-91c4-604eb8b6e9ee.png">

anthonygalea replied:
For determining whether to run automated image comparison on a snippet, what do you think about adding an opt in the snippet? Something like this:
```clj
(defsnippet ortho-perspective
  ["ortho" "perspective"]
  {:renderer :p3d
   :automated-comparison? false}
```
with default `true`. Then we could get rid of the function `skip-automated-compare?`.

nbeloglazov replied:
> No, just checked to be sure, both clj and cljs look like this:
Thank you for providing the screenshot. I was wrong.

> For determining whether to run automated image comparison on a snippet, what do you think about adding an opt in the snippet?
That sounds good. Better than having a whitelist in a separate file. One note though that it would be better to have the field "false" by default. Maybe `:skip-automated-compare? true`. The reason is that usually options are off by default and you turn then on by explicitly setting it. It's just slightly easier to parse mentally.

nbeloglazov commented on test/clj/quil/snippet.clj:
```diff
@@ -14,9 +16,42 @@
+               "time-and-date"} snippet-name))
+
+(defn- compare-images [expected actual difference]
+  (let [{:keys [err]} (sh/sh "compare" "-metric" "mae" expected actual difference)
```
Can you add comment here explaining what is "compare" and what is "mae"? Is that a command line tool that needs to be installed in order to run test?

anthonygalea replied:
Added link to imagemagick docs. To be honest I'm not sure that `mae` is what we should be using. Saw it was used in `kamera` and went with that to start. The `threshold` values I chose are also quite arbitrary at this point. The idea is to get more experience with these parameters while upgrading p5 to 0.8.0 and processing to 3.5.3.

Don't like that we have to install an additional tool to run tests but didn't find any other alternatives in this space so far.

anthonygalea replied:
done

nbeloglazov commented on test/clj/quil/snippet.clj:
```diff
@@ -14,9 +16,42 @@
+
+(defn- compare-images [expected actual difference]
+  (let [{:keys [err]} (sh/sh "compare" "-metric" "mae" expected actual difference)
+        result        (second (re-find #"\((.*)\)" err))]
```
What structure "err" has that you can extract and parse clojure from it?

anthonygalea replied:
`err` is a string with something like this: `23.45 (0.115)`. The second number is what I'm extracting.

anthonygalea replied:
done

nbeloglazov commented on test/clj/quil/snippet.clj:
```diff
@@ -14,9 +16,42 @@
+               "save"
+               "time-and-date"} snippet-name))
+
+(defn- compare-images [expected actual difference]
```
Please add comment to this function explaining what arguments it expects and what it returns. As I understand `difference` is a name of a file and it writes diff image into that file. That's not very clear from just reading function and its implementation.

anthonygalea replied:
done

nbeloglazov commented on test/clj/quil/snippet.clj:
```diff
@@ -14,9 +16,42 @@
+(defn assert-unchanged-snippet-output [name]
+  (when-not (skip-automated-compare? name)
+    (let [n          (str "snippet-snapshots/clj/" name)
+          _          (q/save (str "resources/" n "-actual.png"))
```
Can you add -actual.png into .gitignore so that they don't pollute git status when running test? Same for difference.

anthonygalea replied:
done

nbeloglazov commented on test/clj/quil/snippet.clj:
```diff
@@ -1,10 +1,12 @@
+            [clojure.test :as t]
+            [clojure.java.shell :as sh]
+            [clojure.string :as string]
+            [etaoin.api :as e]
```
Can we use something longer than `e` etaoin? Because `e/go` is really generic function used below and it's hard to understand where it comes from. I'd prefer having full `etaoin` to make keep it explicit.

anthonygalea replied:
done

nbeloglazov commented on test/clj/quil/snippet.clj:
```diff
@@ -81,3 +115,28 @@
+  all-cljs-snippets-produce-expected-output
+  (let [driver (e/firefox)
+        threshold 0.0001]
+    (e/go driver (str "http://localhost:3000/test.html"))
```
No need to wrap in (str) here.

anthonygalea replied:
done

nbeloglazov commented on test/clj/quil/snippet.clj:
```diff
@@ -81,3 +115,28 @@
+    (e/go driver (str "http://localhost:3000/test.html"))
+    (let [elements (e/query-all driver {:tag :option})
+          names    (doall (map #(e/get-element-text-el driver %) elements))
+          c        (count elements)]
```
I don't think you need to alias `(count elements)` as c given that you use `c` only once. Maybe just do `(dotimes [i (count elements)] ...)`?

anthonygalea replied:
done

anthonygalea commented:
Created separate `ortho` and `perspective` snippets, and cleaned up snapshots we don't need. Next is trying this out with new processing and p5 deps. Is there anything I'm missing that prevents this pull request to be merged? Don't want it to get too big.
